src/snowflake/connector/cpp/ArrowIterator/CArrowTableIterator.cpp

```diff
@@ -154,45 +188,45 @@ std::shared_ptr<ReturnVal> CArrowTableIterator::next()
   }
 }
 
-arrow::Status CArrowTableIterator::replaceColumn(
-    const unsigned int batchIdx,
-    const int colIdx,
+void CArrowTableIterator::replaceColumn(
+    const unsigned int batchIdx, const int colIdx,
     const std::shared_ptr<arrow::Field>& newField,
-    const std::shared_ptr<arrow::Array>& newColumn)
+    const std::shared_ptr<arrow::Array>& newColumn,
+    std::vector<std::shared_ptr<arrow::Field>>& futureFields,
+    std::vector<std::shared_ptr<arrow::Array>>& futureColumns,
+    bool& needsRebuild)
 {
   // replace the targeted column
-  std::shared_ptr<arrow::RecordBatch> currentBatch = (*m_cRecordBatches)[batchIdx];
-  arrow::Status ret = currentBatch->AddColumn(colIdx+1, newField, newColumn, &currentBatch);
-  if(!ret.ok())
+  if (needsRebuild == false)
```

Comment:
It looks like we will do the replace anyway. Maybe no need for this judge?

Reply:
Okay, so the idea behind this if then clause is the following.

We have already existing batches that we can turn into a table for `pyarrow`, but if the batches have some data in them that needs modification then we do it at this state.
So `replace_column` is called every time we do one of these modifications, so if this is the first time we are replacing a column then we need to make a deep copy of the the chunk first and then modify that copy and at the end replace the whole chunk.
Making deep copies should not be done in 2 cases. 1. when we have already done it, or 2. if we don't need to modify any of the columns. So this if then clause does it only the one time.

Reply:
Gotcha, thanks for the clarification!